File: servitec_app/lib/core/utils/contact_info_filter.dart
/// Detects contact info or off-platform bypass attempts in chat messages.
///
/// Client-side first line of defense. Server-side enforcement (Cloud Function
/// on message create + Firestore rules) must replicate this — a determined
/// user can patch the client. Tracked in 5-29.md section 4.
class ContactInfoFilter {
  ContactInfoFilter._();

  static final RegExp _email = RegExp(
    r'[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}',
  );

  // Sequences with 8+ digits (allows spaces, dashes, dots, parens).
  // Catches Mexican mobile/landline regardless of country-code prefix.
  static final RegExp _phone = RegExp(
    r'(?:\+?\d[\s\-.()]*){8,}',
  );

  static final RegExp _whatsappUrl = RegExp(
    r'(?:wa\.me|api\.whatsapp\.com|chat\.whatsapp\.com)',
    caseSensitive: false,
  );

  static final RegExp _telegramUrl = RegExp(
    r'(?:t\.me|telegram\.me)',
    caseSensitive: false,
  );

  // Common Spanish bypass phrases. Word boundaries kept loose because users
  // frequently misspell ("wasap", "guasap", "wapp").
  static final RegExp _bypassKeywords = RegExp(
    r'\b(?:whats?\s*app|wha?tsa?pp|wasap|guasap|wapp|telegram|messenger|'
    r'fuera\s+de\s+la\s+app|por\s+fuera|afuera\s+de\s+la\s+app|'
    r'transferencia|deposito\s+directo|pagame\s+por\s+fuera|'
    r'sin\s+la\s+app|sin\s+servitec)\b',
    caseSensitive: false,
  );

  /// Returns `null` if the message is clean, otherwise a [ContactViolation].
  static ContactViolation? scan(String text) {
    if (text.trim().isEmpty) return null;

    if (_email.hasMatch(text)) {
      return const ContactViolation(
        reason: ContactViolationReason.email,
        message:
            'No compartas correos electrónicos en el chat. Mantén la comunicación dentro de ServiTec.',
      );
    }

    if (_whatsappUrl.hasMatch(text) || _telegramUrl.hasMatch(text)) {
      return const ContactViolation(
        reason: ContactViolationReason.externalLink,
        message:
            'No se permiten enlaces a WhatsApp o Telegram. Continúa la conversación aquí.',
      );
    }

    if (_phone.hasMatch(text)) {
      return const ContactViolation(
        reason: ContactViolationReason.phone,
        message:
            'No compartas números telefónicos. ServiTec habilitará el contacto cuando inicie el servicio.',
      );
    }

    if (_bypassKeywords.hasMatch(text)) {
      return const ContactViolation(
        reason: ContactViolationReason.bypassKeyword,
        message:
            'Las transacciones y la comunicación deben mantenerse dentro de ServiTec para tu protección.',
      );
    }

    return null;
  }
}

enum ContactViolationReason { phone, email, externalLink, bypassKeyword }

class ContactViolation {
  final ContactViolationReason reason;
  final String message;

  const ContactViolation({required this.reason, required this.message});
}

File: functions/src/lib/contact-info-filter.ts
/**
 * Server-side port of `servitec_app/lib/core/utils/contact_info_filter.dart`.
 *
 * The Dart version is a UX affordance (it warns before sending); THIS is the
 * enforcement point — a patched client can skip the former but not the latter.
 * Keep the two in sync: same patterns, same precedence, same Spanish copy, so
 * a user sees identical wording whichever layer catches them.
 *
 * Precedence matters: `wa.me/525512345678` must be reported as an external
 * link, not as a phone number, so the URL patterns are tested before the
 * digit-sequence pattern.
 */

export type ContactViolationReason =
  | 'phone'
  | 'email'
  | 'externalLink'
  | 'bypassKeyword';

export interface ContactViolation {
  reason: ContactViolationReason;
  /** User-facing Spanish explanation, shown in place of the blocked message. */
  message: string;
}

const EMAIL = /[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/;

// Sequences of 8+ digits, tolerating spaces, dashes, dots and parens.
// Catches Mexican mobile/landline numbers with or without a country code.
const PHONE = /(?:\+?\d[\s\-.()]*){8,}/;

const WHATSAPP_URL = /(?:wa\.me|api\.whatsapp\.com|chat\.whatsapp\.com)/i;

const TELEGRAM_URL = /(?:t\.me|telegram\.me)/i;

// Common Spanish bypass phrases. Boundaries stay loose because users
// frequently misspell ("wasap", "guasap", "wapp").
const BYPASS_KEYWORDS = new RegExp(
  '\\b(?:whats?\\s*app|wha?tsa?pp|wasap|guasap|wapp|telegram|messenger|' +
    'fuera\\s+de\\s+la\\s+app|por\\s+fuera|afuera\\s+de\\s+la\\s+app|' +
    'transferencia|deposito\\s+directo|pagame\\s+por\\s+fuera|' +
    'sin\\s+la\\s+app|sin\\s+servitec)\\b',
  'i',
);

/** Returns `null` when the message is clean, otherwise the violation. */
export function scan(text: string): ContactViolation | null {
  if (!text || text.trim().length === 0) return null;

  if (EMAIL.test(text)) {
    return {
      reason: 'email',
      message:
        'No compartas correos electrónicos en el chat. Mantén la comunicación dentro de ServiTec.',
    };
  }

  if (WHATSAPP_URL.test(text) || TELEGRAM_URL.test(text)) {
    return {
      reason: 'externalLink',
      message:
        'No se permiten enlaces a WhatsApp o Telegram. Continúa la conversación aquí.',
    };
  }

  if (PHONE.test(text)) {
    return {
      reason: 'phone',
      message:
        'No compartas números telefónicos. ServiTec habilitará el contacto cuando inicie el servicio.',
    };
  }

  if (BYPASS_KEYWORDS.test(text)) {
    return {
      reason: 'bypassKeyword',
      message:
        'Las transacciones y la comunicación deben mantenerse dentro de ServiTec para tu protección.',
    };
  }

  return null;
}
